File: frontend/debug/deduplicator.js
/**
 * deduplicator.js — Phase 1
 *
 * Client-side deduplication filter.
 *
 * Mirrors the backend logic in utils/deduplicator.py:
 *   Key:    error_code + position rounded to nearest 1mm
 *   Window: 10 seconds (default, matches settings.json)
 *
 * The server also deduplicates, so this is a fast first-pass filter
 * that avoids unnecessary network requests when the same error fires
 * multiple times in rapid succession.
 */

const Deduplicator = (() => {
  // { key: timestamp_ms }
  const _seen = new Map();

  const _WINDOW_MS = 10_000;  // 10 seconds

  // ── Public API ─────────────────────────────────────────────────────────

  /**
   * Returns true if this error+position has been seen within the window.
   * Registers the event if it is new.
   *
   * @param {string} errorCode
   * @param {{ x: number, y: number, z: number }} position
   * @param {number} [windowMs]  override default 10 000 ms
   */
  function isDuplicate(errorCode, position, windowMs = _WINDOW_MS) {
    _evictExpired(windowMs);

    const key = _makeKey(errorCode, position);
    if (_seen.has(key)) return true;

    _seen.set(key, Date.now());
    return false;
  }

  /** Clear all state (useful for testing). */
  function clear() {
    _seen.clear();
  }

  // ── Helpers ────────────────────────────────────────────────────────────

  function _makeKey(errorCode, position) {
    const x = Math.round(Number(position?.x) || 0);
    const y = Math.round(Number(position?.y) || 0);
    const z = Math.round(Number(position?.z) || 0);
    return `${errorCode}:${x},${y},${z}`;
  }

  function _evictExpired(windowMs) {
    const cutoff = Date.now() - windowMs;
    for (const [key, ts] of _seen.entries()) {
      if (ts < cutoff) _seen.delete(key);
    }
  }

  return { isDuplicate, clear };
})();

File: utils/deduplicator.py
"""
Error deduplication logic for the CNC assistant.

Dedup key: error_code + machine position rounded to 1mm.
Window: 10 seconds (configurable).

If the same error fires multiple times within the window, only the first
is passed through — the rest are silently dropped to prevent API spam.
"""

import hashlib
import time
from typing import Optional

# { hash_key: timestamp_of_first_occurrence }
_seen: dict[str, float] = {}


def _round_position(pos: dict) -> tuple:
    """Round x/y/z to nearest mm for dedup purposes."""
    return (
        round(float(pos.get("x", 0))),
        round(float(pos.get("y", 0))),
        round(float(pos.get("z", 0))),
    )


def _make_key(error_code: str, position: dict) -> str:
    rounded = _round_position(position)
    raw = f"{error_code}:{rounded[0]},{rounded[1]},{rounded[2]}"
    return hashlib.sha1(raw.encode()).hexdigest()


def is_duplicate(error_code: str, position: dict, window_seconds: float = 10.0) -> bool:
    """
    Returns True if this error+position was already seen within the window.
    Registers the event if it is new.
    """
    _evict_expired(window_seconds)
    key = _make_key(error_code, position)
    now = time.monotonic()

    if key in _seen:
        return True

    _seen[key] = now
    return False


def _evict_expired(window_seconds: float):
    now = time.monotonic()
    expired = [k for k, t in _seen.items() if now - t > window_seconds]
    for k in expired:
        del _seen[k]


def clear():
    """Clear all dedup state (useful for testing)."""
    _seen.clear()
